### PR TITLE
Update CI to support python 3.12

### PR DESCRIPTION
**Pull request by kAIto47802**
## Motivation

Update CI checking to support python 3.12.

## Description of the changes

- Add Python 3.12 checking to the CI.
- Remove the Python 3.12 checking for components incompatible with Python 3.12.

TensorFlow and MXNet will not be tested after this PR.

The reasons why I removed their Python 3.12 checking are as follows:

- TensorFlow: [Migrating from Estimator to Kras APIs is recommended](https://www.tensorflow.org/guide/migrate/migrating_estimator).
- MXNet: Not compatible with Python 3.12.
   - MXNet  development ended and [moved into the Apache Attic in September 2023](https://attic.apache.org/projects/mxnet.html).
   - The latest version of MXNet, 1.9.1, is not compatible with NumPy versions 1.24 or higher.
      - This is because the code uses `np.bool`, which has been deprecated and was removed in NumPy 1.24 [^1].
   - The version of NumPy that supports Python 3.12 is 1.26.0 or higher.
      - This is because Python 3.12 removed `distutils`, which earlier versions of NumPy relied on.

[^1]: https://numpy.org/devdocs/release/1.24.0-notes.html#expired-deprecations

### CODE REVIEW

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
Could you apply a similar change for MXNet as well?

**Review by c-bata — COMMENTED** [review]
Thank you for your pull request.

I’m wondering if we should consider deprecating the MXNet integration, as its development has been discontinued.
https://mxnet.apache.org/versions/1.9.1/

File: .github/workflows/checks-and-tests.yml
Comment (c-bata) [inline .github/workflows/checks-and-tests.yml:222]: I'm wondering if we could remove the unit tests and CI settings for testing `tensorflow` integration and `mxnet` integration since these features are deprecated.

Comment (kAIto47802) [reply .github/workflows/checks-and-tests.yml:222]: Thank you for your comment. I have added `deprecated_class` decorator to MXNet integration.

**Review by c-bata — APPROVED** [review]
LGTM after CI passes. I’ve also created a follow-up PR (#175) for the PyPI badges in the README and package metadata.

**Comment by c-bata** [comment]
@nabenabe0928 Please merge this PR after you approve.

**Review by nabenabe0928 — APPROVED** [review]
LGTM!

**Pull request merged**
